src/server_manager/web_app/app.ts

```diff
@@ -200,136 +209,218 @@ export class App {
-    this.manualServerRepository.listServers().then((manualServers) => {
-      // Show any manual servers if they exist.
-      if (manualServers.length > 0) {
-        this.showManualServerIfHealthy(manualServers[0]);
-        return;
-      }
+    // Load display servers and associate them with managed and manual servers.
+    this.displayServerRepository.listServers().then((displayServers) => {
+      this.appRoot.serverList = displayServers;
+
+      this.manualServerRepository.listServers().then((manualServers) => {
+        this.syncServersToDisplay(manualServers).then(() => {
+          this.maybeShowLastDisplayedServer();
+        });
+      });
 
-      // User has no manual servers - check if they are logged into DigitalOcean.
       const accessToken = this.digitalOceanTokenManager.getStoredToken();
       if (accessToken) {
-        this.enterDigitalOceanMode(accessToken);
-        return;
+        this.enterDigitalOceanMode(accessToken).then((managedServers) => {
+          this.syncServersToDisplay(managedServers).then(() => {
+            this.maybeShowLastDisplayedServer();
+          });
+        });
       }
+    });
+    this.showIntro();
+  }
 
-      // User has no manual servers or DigitalOcean token.
-      this.showIntro();
+  private async syncServersToDisplay(servers: server.Server[]) {
+    for (const server of servers) {
+      await this.syncServerToDisplay(server);
+    }
+  }
+
+  // Syncs the locally persisted server metadata for `server`. Creates a DisplayServer for `server`
+  // if one is not found in storage.
+  private async syncServerToDisplay(server: server.Server): Promise<DisplayServer> {
+    const displayServerId = server.getManagementApiUrl();
+    let displayServer = this.displayServerRepository.findServer(displayServerId);
+    if (!displayServer) {
+      console.warn(`Could not find managed display server with ID ${displayServerId}`);
+      const isHealthy = await server.isHealthy().catch((e) => false);
+      displayServer = {
+        id: displayServerId,
+        name: isHealthy ? server.getName() : server.getHostname(),
+        isManaged: isManagedServer(server)
+      };
+      this.displayServerRepository.addServer(displayServer);
+      this.syncDisplayServersToUi();
+    } else {
+      // We may need to update the stored display server if it was persisted when the server was not
+      // healthy.
+      try {
+        const remoteServerName = server.getName();
+        if (displayServer.name !== remoteServerName) {
+          displayServer.name = remoteServerName;
+          this.removeServerFromDisplay(displayServer);
+          this.displayServerRepository.addServer(displayServer);
+          this.syncDisplayServersToUi();
+        }
+      } catch (e) {
+        // Ignore, we may not have the server config yet.
+      }
+    }
+    this.serverDisplayTable[displayServer.id] = {displayServer, server};
+    return displayServer;
+  }
+
+  private syncDisplayServersToUi() {
```

Comment:
This function needs a comment, especially the part about triggering a Polymer mutation - does Polymer not detect changes in the object? Right now you call it every time you fetch a server's info which seems wasteful, albeit reliable.

Reply:
Excellent.